Login Page

banciur opened the pull request.

banciur committed:
Initial commit of 3 column layout

banciur committed:
Get rid of one of two unique "key" prop warnings

recmo reviewed (commented):
I think the deciding factor should be what usage looks like. I did an experiment comparing children to props.

### Using `children`

```js
const ex = (
  <ThreeColumn>
    <div>
      <h1>Hello1</h1>
      <p>Left column</p>
    </div>
    <div>
      <h1>Malcom</h1>
      <p>Middle column</p>
    </div>
    <div>
      <h2>Wilders</h2>
      <p>Right column</p>
    </div>
  </ThreeColumn>
);
```

Pro: Simple
Con: It's not clear that you need to have exactly three children. What happens if you don't?

### Using properties

```js
const ex = (
  <ThreeColumn
    left={
      <div>
        <h1>Hello1</h1>
        <p>Left column</p>
      </div>
    }
    middle={
      <div>
        <h1>Malcom</h1>
        <p>Middle column</p>
      </div>
    }
    right={
      <div>
        <h2>Wilders</h2>
        <p>Right column</p>
      </div>
    }
  />
);
```

Pro: Immediately obvious semantics.
Con: More verbose syntax

banciur commented:
@Recmo yes and I also had problem with those small helper functions 
`const entryPage = components => () => <LayoutEntry>{components}</LayoutEntry>;`
I wanted to use them like this:
`entryPage([<Login  />, <OtherComponent />, <OtherComponent />]),` or `entryPage(<Login  />),` but it was also giving me missing key prop warning but once after page was displayed for first time.

I lost to much time here -  I'll implement properties approach it will be fast and easy.

recmo commented on src/routes.js:
```diff
@@ -12,13 +12,14 @@ import NotFound from './views/NotFound';
 
-const entryPage = View => () => <LayoutEntry><View /></LayoutEntry>;
+// TODO: this call gives react warning about "unique key propety"
+const entryPage = components => () => <LayoutEntry>{components}</LayoutEntry>;
```
This is a weird definition. I'm not sure how it works to pass components like that.

recmo commented on src/routes.js:
```diff
@@ -12,13 +12,14 @@ import NotFound from './views/NotFound';
   '/': page(QuotedStartups),
   '/test': page(Test),
-  '/login': entryPage(Login),
+  '/login': entryPage([<Login key="12" />, <div>test</div>, 'test']),
```
I don't think it makes sense to define the columns in the routing file. There should simply be a `entryPage(LoginView)`. How the `LoginView` looks (layout, columns, etc) is up to the implementation.

recmo commented on src/views/LayoutEntry.js:
```diff
@@ -1,15 +1,25 @@
-  </div>
-  <Footer />
-</div>);
+const LayoutEntry = ({ children }) =>
```
This component has two functions:

* It wraps the children in a header and footer.
* It (tries to) align the children in three columns.

I think these are two separate functions that should not be mixed in one component. This component should focus purely on the header/footer (like the `Layout` component it is named after).

The layout should go directly into the `LoginView`. We don't need a special component for a three column layout, react-flexbox-grid makes it easy enough.

(Now, if it turns out that we are repeating the same layout in many views, we can consider factoring it out. But layouts are so tightly integrated with their views that this will be hard, and should only be done if there is a significant benefit, like a big reduction in repeated code. The important thing is, we don't do this now.)

recmo reviewed (changes requested):
I don't think this is the right approach. Let's discuss.

banciur committed:
Initial markup for LoginPage

banciur committed:
LoginPage first iteration

banciur committed:
LoginPage css fixes

banciur committed:
Header_type_1 fix color names and change button type

banciur committed:
Header_type_1 center button in its column

banciur committed:
Add action to navigate to registration page to header button

banciur committed:
Add event.preventDefault() to click handling on register button

banciur committed:
Change right border on header logo to correct one

banciur committed:
Some cleaning

Replaced Footer with Footer_temp
Removed commented functions and imports from routes.js
Used some colors from imports in few places

banciur commented on src/pages/LoginPage.js:
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
```
@Recmo I would like to have your opinion on this component. Yesterday we talked that maybe it makes sense to implement pages as one component without using any layouts. And later when we observe to much markup repetition we will create some generic layouts. This is outcome of first login page.

recmo replied:
Looks great to me! Clear, straightforward.

recmo commented on src/pages/LoginPage.js:
```diff
@@ -0,0 +1,43 @@
+        <Grid fluid>
+          <Row>
+            <Col lgOffset={1} lg={10}>
+              <Row>
```
There should not be a `Row` inside a `Col`. Instead, following Ola's layout, there should by three `Col` instances, all with `lg=3`, the first one will have an offset.

banciur replied:
If we do 1 column offset then 3 x 3 columns we will end with 10 columns. So on left side there will be one column and two on right and content it won't be centred.
Nesting grids is feature of flebox layout described in doc. Maybe (didn't try it) mixing fixed size columns and auto width feature would work. Something like
```
<Col lgOffset={1} lg>
<Col lg>
<Col lg>
<Col lg={1}>
```
but then we would end with extra empty column.

recmo replied:
Ah, I get it. You're right. The way Ola is using the 12 grid, effectively turns it into a 10 grid. And then you can't have three columns. Let me talk to her.

recmo commented on src/pages/LoginPage.js:
```diff
@@ -0,0 +1,43 @@
+            <Col lgOffset={1} lg={10}>
+              <Row>
+                <Col lg={4}>
+                  <div className={styles.columnArea}>
```
Can we remove the `div` and apply the style directly to the `Col` instance?

banciur replied:
Sadly no, Col component is rendered with padding that takes care for spacing between columns.
And .column-area uses another padding to put gap between border of white area and inside elements.

banciur commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,54 @@
+            </Col>
+            <Col lg={3} className={styles.createButton}>
+              <RoundHeaderButton
+                onClick={(event) => {
```
@Recmo Here I'm wondering about correct way to implement it. I'm hard coding `/register` address into component which is not best idea.  
We already have [PredefinedLinks](https://github.com/Neufund/generic-ico/blob/master/src/components/PredefinedLinks.js) component which renders <a> elements. Still surrounding <button> with <a> element is mistake. 

So now I ended in having `/register` path in two places. Should I abstract those targetTo to calls and reuse it across Header and PredefinedLinks?

recmo replied:
Hmm.. Either you use `RegisterLink` and apply a `className` to it to make it look like the round header button, or we keep the current implementation. Both approaches feel about equally messy to me.

Alternatively, we can modify `PredifinedLink` to take a component as a property, then it would look like this:

```
<RegisterLink button={RoundHeaderButton}>{i18n('…')}</RegisterLink>
```

banciur replied:
I'm thinking maybe lets move it as TODO and do it with another PR. I think that header will change anyway.

recmo commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,54 @@
+            <Col lg={3} className={styles.createButton}>
+              <RoundHeaderButton
+                onClick={(event) => {
+                  event.preventDefault();
```
Do we need `preventDefault` here? I thought it was only required for `<a>` tags. (And then `Link` should take care of it for us.)

banciur replied:
Yes we don't need it here. It's my jQuery habits kicked in :)

recmo commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,54 @@
+                  event.preventDefault();
+                  targetTo('/register');
+                }}
+                style={{ display: 'inline-block' }}
```
aren't buttons `inline-block` by default? I think this should move to the CSS.

banciur replied:
Adding @Moejoe90  as this might be interesting also to you. 
Our current buttons have `display: block` by default. I wasn't sure if it was something you planned when implementing those elements.
If we want to copy existing patterns then for example material-ui buttons are indeed `inline-block`. I can create PR for this. 

Setting this through styles at our current implementation is not possible. If I would do this
`<RoundHeaderButton className={styles.ownStyle}>...</RoundHeaderButton>`
Rendered element would receive styles only from my `ownStyle` class.  

Anyway if we change implementation that button would receive styles from both style-sheets we would have `display` property defined on same level in two different files. Outcome which one will be used by browser would be not so obvious - css path length and rule order would define that. Or `!important` keyword. 

I don't see straightforward elegant solution here.

recmo replied:
I meant it should be default for a button. If you want to override, I would do it with a `style` prop like you do. But in this case I think it should be default behaviour from the `RoundHeaderButton`.

banciur replied:
I did it in separate PR to so now its no longer in code

banciur committed:
Header.js remove unnecessary preventDefault() call

banciur committed:
Removed unnecessary style override

banciur committed:
Add missing <Grid> component in LoginPage

banciur committed:
Removed unused HeaderInitial and LayoutEntry components

banciur committed:
Adds todo comment to Header_type_1 how to solve hardcoded link

banciur committed:
Remove unused scss variable from Header_type_1.scss

banciur committed:
Move shadow colors declarations to colors.scss

banciur committed:
Rename main react component variable in Header_type_1

banciur committed:
Removed unused style import from common.scss

recmo commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,50 @@
+import styles from './Header_type_1.scss';
+import common from '../styles/common.scss';
+
+const HeaderType1Component = ({ i18n, targetTo }) =>
```
Component name is not identical to file name. In this case the filename *needs* to be `HeaderType1.js`.

@Moejoe90 @banciur after you do a PR, you need to also go through the PR yourself and check for simple stuff like this.

banciur replied:
Guilty as charged. I wont make more of that kind of mistakes.

recmo commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,50 @@
+            <Col lg={3} className={styles.createButton}>
+              { /* TODO: hardcoding link is not best idea. There is discussion how this can be
+                changed - https://github.com/Neufund/generic-ico/pull/21#discussion_r121947303 */}
+              <RoundHeaderButton onClick={() => targetTo('/register')}>
```
This logic should not be in the component. You can make an `onCreateAccountClick` prop.

recmo commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,50 @@
+
+export default connect(
+  state => ({ i18n: getTranslator(state) }),
+  dispatch => ({ targetTo: path => dispatch(routeTo(path)) })
```
And here you can define the `onCreateAccountClick` as a `routeTo` with the path.

banciur replied:
Thank you this is much better approach.

recmo commented on src/components/Header_type_1.js:
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
```
Can we come up with a more descriptive name for the header, like `HeaderUnauthenticated` or `HeaderPreAuthenticated`

banciur committed:
Rename Header_type_1 to HeaderUnauthenticated

banciur committed:
Remove hardcoded link from HeaderUnauthenticatedComponent

recmo reviewed (approved):
LGTM!
